### PR TITLE
Disallow same state transitions

### PR DESCRIPTION
**Pull request by menanteau**
Update the state_matrix in states.py to disallow same state transitions
Update salpylib to send SAL__CMD_NOPERM value when transition is not allowed

### CODE REVIEW

File: python/salpytools/salpylib.py [hunk 1]
```diff
@@ -235,6 +235,8 @@ def subscribe(self):
         self.mgr_ackCommand = getattr(self.mgr, 'ackCommand_{}'.format(self.command))
         # And the code for the SAL__CMD_COMPLETE from the library itself
         self.SAL__CMD_COMPLETE = SALPY_lib.SAL__CMD_COMPLETE
+        self.SAL__CMD_FAILED = SALPY_lib.SAL__CMD_FAILED
+        self.SAL__CMD_NOPERM = SALPY_lib.SAL__CMD_NOPERM
```
Comment (ktlim) [inline python/salpytools/salpylib.py]: It seems a little strange to copy all these into the instance when they're constant and won't be changed.  Why not just use them directly from `SALPY_lib`?  (Also, you have `SALPY_lib` as a `subscribe()` method variable, as an instance variable set in `subscribe()` and as an instance variable set in `__init__()`.  There doesn't seem to be a reason for handling this three different ways.)

File: python/salpytools/salpylib.py [hunk 2]
```diff
@@ -283,8 +285,10 @@ def reply_to_transition(self, cmdId):
             LOGGER.info(
                 "WARNING: INVALID TRANSITION from {} --> {}".format(self.State.current_state,
                                                                     self.next_state))
-            self.State.send_logEvent('rejectedCommand', rejected_state=self.COMMAND)
-
+            # Send the ACK
+            self.mgr_ackCommand(cmdId, self.SAL__CMD_NOPERM, 0, "INVALID TRANSITION")
+            # The rejectCommand seems to be deprectaded.
```
Comment (ktlim) [inline python/salpytools/salpylib.py]: There doesn't seem to be any reason to keep this code, even commented out.

File: python/salpytools/salpylib.py [hunk 3]
```diff
@@ -283,8 +285,10 @@ def reply_to_transition(self, cmdId):
             LOGGER.info(
                 "WARNING: INVALID TRANSITION from {} --> {}".format(self.State.current_state,
                                                                     self.next_state))
-            self.State.send_logEvent('rejectedCommand', rejected_state=self.COMMAND)
-
+            # Send the ACK
+            self.mgr_ackCommand(cmdId, self.SAL__CMD_NOPERM, 0, "INVALID TRANSITION")
```
Comment (ktlim) [inline python/salpytools/salpylib.py]: I believe it's best practice to include the source and destination states in the message, just like the log message above.

File: python/salpytools/salpylib.py [hunk 4]
```diff
@@ -465,8 +469,10 @@ def __init__(self, Device, sleeptime=1, timeout=5, threadID=1):
         LOGGER.info("Loading Device: {}".format(self.Device))
         # Load SALPY_lib into the class
         self.SALPY_lib = load_SALPYlib(self.Device)
-        # And the code for the SAL__CMD_COMPLETE from the library itself
+        # Shorcuts the code for some the SAL__CMD_XXXXX from the library itself
```
Comment (ktlim) [inline python/salpytools/salpylib.py]: Saving 10 characters on use by adding two lines and possible spelling errors here does not seem like a great tradeoff.

File: python/salpytools/states.py [hunk 5]
```diff
@@ -74,11 +74,11 @@
...
 
-# Set up same state transitions as allowed
-state_matrix[0][0] = True
-state_matrix[1][1] = True
-state_matrix[2][2] = True
-state_matrix[3][3] = True
-state_matrix[4][4] = True
-state_matrix[5][5] = True
-state_matrix[6][6] = True
+# Disallow same state transitions
```
Comment (ktlim) [inline python/salpytools/states.py:77]: Seems like "Disallowing" should involve setting to False, not commented out setting to True.

Comment (menanteau) [reply python/salpytools/states.py:77]: The `state_matrix` matrix context are all set to False by default. So they need to be change to `True`

File: python/salpytools/salpylib.py (hunk 1 above)
Comment (menanteau) [reply python/salpytools/salpylib.py]: Yes, that was bothering me too and should I have simple use `SALPY_lib` and avoid these shortcut. Thanks.

File: python/salpytools/salpylib.py (hunk 3 above)
Comment (menanteau) [reply python/salpytools/salpylib.py]: Good suggestion, I'll add that too.

File: python/salpytools/salpylib.py (hunk 2 above)
Comment (menanteau) [reply python/salpytools/salpylib.py]: I think you are right, but Michael still wanted to discuss this next Tuesday. I'll remove it anyway. We can also resurrect it later.

**Pull request merged**